Add a time limit for optimizer rules

maciejgrzybek opened the pull request:
This is to ensure the query execution finishes even in case of non-converging optimizer rule.
This can be helpful when in testing as well, in case build is stuck because of bogus rule.

The time limit is not strictly enforced because planning time is recorded per rule only approximately.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java:
```diff
@@ -134,6 +157,12 @@ private boolean exploreNode(int group, Context context)
+    // It's not a hard limit due to inaccuracy of time registered in ruleStats
+    private boolean isTimeLimitExhausted()
+    {
+        return stats.getApproxTotalTimeInMillis() >= timeLimitInMillis;
```
Instead of measuring time by relying on what's tracked in `stats`, just record the start time when `optimize` is called (store it in the Context object) and check against that.

maciejgrzybek replied:
Addressed in 910e05b06a12ea9dc39f466b060408ed962304ef.

martint commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/TestIterativeOptimizer.java:
```diff
@@ -0,0 +1,89 @@
+        @Override
+        public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
+        {
+            return Optional.of(node);
```
This may not be a good way to produce a non-converging rule in the long term. I could see us updating the optimizer to ignore the result of the transformation if it's exactly the same object that was passed in (I though I had added this already, but maybe it was in some prototype code I was working on).

maciejgrzybek replied:
It was working for now but I addressed your concerns in 4f213fffc92cc6eedd8a6dfb0b25da5e4c4ba443.

martint commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/TestIterativeOptimizer.java:
```diff
@@ -0,0 +1,89 @@
+            });
+        }
+        catch (PrestoException ex) {
+            assertEquals(ex.getErrorCode(), OPTIMIZER_TIMEOUT.toErrorCode());
```
This test will succeed if the rule does converge, so you need to call `fail()` if it doesn't trigger the optimizer timeout.

maciejgrzybek commented:
Back to you, @martint.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java:
```diff
@@ -46,11 +52,24 @@ public IterativeOptimizer(StatsRecorder stats, Set<Rule> rules)
 
     public IterativeOptimizer(StatsRecorder stats, List<PlanOptimizer> legacyRules, Set<Rule> newRules)
     {
+        this(stats, legacyRules, newRules, 5000);
```
I thought I had made a comment about making the default timeout configurable and optional, but I can't find it now.

A timeout of 5s is not necessarily a good general setting. Some optimizers might need to access external metadata and could be sensitive to connector choice and other environmental properties.

maciejgrzybek replied:
I don't remember such a comment. But I agree with it completely.
Will provide a property to configure it. In case it's not specified, timeout is not enforced.

maciejgrzybek commented:
Back to you, @martint.

martint commented on presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:
```diff
@@ -279,6 +280,15 @@ public SystemSessionProperties(
                         false),
+                new PropertyMetadata<>(
+                        ITERATIVE_OPTIMIZER_TIMEOUT,
+                        "Timeout for plan optimization in iterative optimizer (0 - timeout disabled)",
```
Hmm... using 0 to mean "disabled" is weird and inconsistent. Unfortunately, I realized there's no way to set a session property to "missing" (unlike configuration options). In light of this, I think it'd be better to not give 0 any special meaning and set the default to a very large number that should never happen in practice (in the order of several minutes. Say, somewhere between 1 and 10 minutes).

martint commented on presto-main/etc/config.properties:
```diff
@@ -41,3 +41,4 @@ plugin.bundles=\
 presto.version=testversion
 distributed-joins-enabled=true
 node-scheduler.include-coordinator=true
+experimental.iterative-optimizer-timeout=60s
```
This config file is used for development (when running the presto server from your IDE). How does it relate to what you've described in the commit message?

maciejgrzybek replied:
I thought this would be the example config where you look first when configuring Presto.
But with https://github.com/prestodb/presto/pull/7458#discussion_r105232529, we probably don't need to change that configuration.
I'll address https://github.com/prestodb/presto/pull/7458#discussion_r105232529 and discard this commit.

maciejgrzybek commented:
Changed as you requested. Rebased on master. Ready to merge! :)

Default timeout is now 3 minutes.
10 minutes is too close to already existing Travis timeout (20 mins), 1 min though could be too short for slow network and multiple remote requests to `ConnectorMetadata`.

I also added a check https://github.com/prestodb/presto/pull/7458/files#diff-5eed8b9c2cc6e5f6be8077b8c2328943R179.

martint commented on presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestFeaturesConfig.java:
```diff
@@ -60,6 +63,7 @@ public void testDefaults()
                 .setOptimizeMixedDistinctAggregations(false)
                 .setLegacyOrderBy(false)
                 .setIterativeOptimizerEnabled(false)
+                .setIterativeOptimizerTimeout(new Duration(0, MILLISECONDS))
```
This needs to be update to 3 minutes. I'll make the change before I merge.

martint commented:
Merged, thanks!

maciejgrzybek commented:
@martint I've noticed that I overlooked that when our internal Jenkins reported that. Sorry for confusion!

BTW what's the point of `TestFeaturesConfig.testDefaults` test?
It looks to me that it only tests whether defaults are set as we wrote in the test... It's just a place with duplicates of default settings, isn't it?
